### PR TITLE
US-7.6: Login as a User

### LINKED ISSUE #99: US-7.6: Login as a User
**User Story Description**

As a User, I want to login so that I can use the Mull application.

**User Acceptance Flow**

1. User opens the app
1. User sees multiple ways of logging in
    a. Through an existing service (Google, Twitter, Facebook, etc)
    b. Using user's email
1. User logs in
1. User sees the home page

**Acceptance Criteria**

- [ ] User is authenticated
- [ ] User cannot access the application without logging in
- [ ] User Acceptance Flow is respected
- [ ] Documentation present where necessary
- [ ] Unit and UI tests are implemented for every component and function
- [ ] There must exist an automated Cypress acceptance test which tests the user acceptance flow
- [ ] Must match [Mockup](create-mockup-link)

### PR DESCRIPTION
**Pull request by bit172**
This PR closes #99.

**Steps to test**
1. Launch the application using `nx serve mull-ui` & `nx serve mull-api`.
2. Navigating to any route except `/login` & `/register` should redirect you to `/login`
3. Create an account if you do not have one to test with.
4. Login with email and pass.
5. In the application tab of the browser developer tools, you should see a cookie called `mullToken`.
6. Closes the browser or tab.
7. Go back to the application.
8. You should be logged in as the same user.
9. Clear the cookies of the browser.
10. Repeat steps 5 to 8 for OAuth login.

### CODE REVIEW

**Review by cindyslittleplanet — COMMENTED** [review]
The pages that were not created yet (map, tools, messages, profile) are still accessible without authentication. 

- [x] User is authenticated
- [ ] User cannot access the application without logging in
- [x] User Acceptance Flow is respected
- [x] Documentation present where necessary
- [x] Unit and UI tests are implemented for every component and function
- [x] There must exist an automated Cypress acceptance test which tests the user acceptance flow
- [x] Must match [Mockup](create-mockup-link)

**Review by r-saba — CHANGES_REQUESTED** [review]
Like @cindyslittleplanet mentioned if a user browses to any undefined route like `http://localhost:4200/map` they see the navigation/header. Can you fix this? Maybe add a 404 page to catch all invalid URL's and make sure the navigation/headers are only visible if the user is logged in

File: apps/mull-ui/src/app/app.tsx [hunk 1]
```diff
@@ -68,19 +94,20 @@ export const App = () => {
...
+                <PrivateRoute path={ROUTES.MY_EVENTS} component={MyEventsPage} />
+                <PrivateRoute exact path={ROUTES.HOME}>
                   <Redirect to={ROUTES.DISCOVER} />
-                </Route>
+                </PrivateRoute>
               </SwipeableRoutes>
             </div>
-          </Route>
+          </PrivateRoute>
+          <Route path="*" component={NotFoundPage} />
```
Comment (r-saba) [inline apps/mull-ui/src/app/app.tsx]: We don't need the path prop by default it will match any route that isn't matched
```suggestion
          <Route component={NotFoundPage} />
```

**Review by r-saba — APPROVED** [review]
LGTM! There's a small refactoring comment I left. No need to re-request if you apply it.

**Acceptance Criteria**

- [x] User is authenticated
- [x] User cannot access the application without logging in
- [x] User Acceptance Flow is respected
- [x] Documentation present where necessary
- [x] Unit and UI tests are implemented for every component and function
- [x] There must exist an automated Cypress acceptance test which tests the user acceptance flow
- [x] Must match [Mockup](create-mockup-link)

**Review by cindyslittleplanet — APPROVED** [review]
All is good now! Good job! 

- [x] User is authenticated
- [x] User cannot access the application without logging in
- [x] User Acceptance Flow is respected
- [x] Documentation present where necessary
- [x] Unit and UI tests are implemented for every component and function
- [x] There must exist an automated Cypress acceptance test which tests the user acceptance flow
- [x] Must match [Mockup](create-mockup-link)

File: apps/mull-api/src/app/auth/auth.service.ts [hunk 2]
```diff
@@ -49,6 +49,7 @@ export class AuthService {
     res.cookie('mullToken', refreshToken, {
       maxAge: 7 * 24 * 3600 * 1000, // 7 days in ms
       secure: environment.production, // Determines if cookie is to be used with HTTPS only
+      httpOnly: true,
```
Comment (cristian-aldea) [inline apps/mull-api/src/app/auth/auth.service.ts:52]: Just a question, won't these two interfere with each other, judging from the `// Determines if cookie is to be used with HTTPS only` comment that was left?

File: apps/mull-ui/src/app/access-token.ts [hunk 3]
```diff
@@ -0,0 +1,9 @@
+let accessToken: string | null = null;
```
Comment (cristian-aldea) [inline apps/mull-ui/src/app/access-token.ts:1]: Should this maybe be part of the user context? Why did you remove it from there

File: apps/mull-ui/src/app/pages/404/not-found.tsx [hunk 4]
```diff
@@ -0,0 +1,11 @@
+import React from 'react';
+
+const NotFoundPage = () => {
```
Comment (cristian-aldea) [inline apps/mull-ui/src/app/pages/404/not-found.tsx]:
```suggestion
const NotFound = () => {
```
Or change the file name to `not-found-page` so that it matches.

File: apps/mull-ui/src/app/app.tsx [hunk 5]
```diff
@@ -39,24 +56,33 @@ export const App = () => {
     return {};
   };
 
+  // TODO: Replace with spinner or loading component
+  if (loading) return <div>Loading...</div>;
+
   return (
-    <UserProvider value={{ userId, setUserId, accessToken, setAccessToken }}>
+    <UserProvider value={{ userId, setUserId }}>
```
Comment (cristian-aldea) [inline apps/mull-ui/src/app/app.tsx:63]: IDK if this was already discussed or resolved, but I still don't get redirected to the login page on existing routes like /map or /asdfasdf. It's good to show the 404 page when you're logged in, but maybe not when you're not.

**Review by cristian-aldea — CHANGES_REQUESTED** [review]
Everything looks great! Good work Jon. Just had some questions as to some of the choices you made here, but won't be blocking over them.

However, tests in mull-ui are failing for me. It might just be on my side but until I get confirmation that it's only me I'll block the PR.

image

File: apps/mull-api/src/app/auth/auth.service.ts (hunk 2 above)
Comment (bit172) [reply apps/mull-api/src/app/auth/auth.service.ts:52]: No, there's a bit of a misconception with the naming of the option. `httpOnly` refers to how and where the cookie is used. This is mainly to allow the cookie to be used only when communicating with the HTTP/S protocols. Meaning it cannot be accessed in javascript. https://owasp.org/www-community/HttpOnly

File: apps/mull-ui/src/app/access-token.ts (hunk 3 above)
Comment (bit172) [reply apps/mull-ui/src/app/access-token.ts:1]: I removed it since it needs to be accessible outside the React Application. Mainly due to the fact that it needs to be used by Apollo to set the authorization header.

File: apps/mull-ui/src/app/pages/404/not-found.tsx (hunk 4 above)
Comment (bit172) [reply apps/mull-ui/src/app/pages/404/not-found.tsx]: I just put the name out of habit. 😅 There doesn't really seem to be a convention for the components in the pages folder.

File: apps/mull-ui/src/app/app.tsx (hunk 5 above)
Comment (bit172) [reply apps/mull-ui/src/app/app.tsx:63]: I will address this issue.

File: apps/mull-ui/src/app/pages/404/not-found.tsx (hunk 4 above)
Comment (cristian-aldea) [reply apps/mull-ui/src/app/pages/404/not-found.tsx]: yeah that's unfortunately, its not a big deal but a fix ticket down the line could address this.

File: apps/mull-ui/src/app/app.tsx (hunk 5 above)
Comment (cristian-aldea) [reply apps/mull-ui/src/app/app.tsx:63]: thanks for addressing it!

**Review by cristian-aldea — APPROVED** [review]
Thanks for addressing my comments, everything looks good.

- [x] User is authenticated
- [x] User cannot access the application without logging in
- [x] User Acceptance Flow is respected
- [x] Documentation present where necessary
- [x] Unit and UI tests are implemented for every component and function
- [x] There must exist an automated Cypress acceptance test which tests the user acceptance flow
